dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java

```diff
@@ -461,9 +461,32 @@ public void handleException(HttpServletRequest req, HttpServletResponse rsp, Thr
     }
 
     @Override
+    @FFDCIgnore(NumberFormatException.class)
     public boolean isCDINeeded() {
-        return WebContainer.getServletContainerSpecLevel() >= WebContainer.SPEC_LEVEL_40 &&
-               provisionerService.getInstalledFeatures().contains("appSecurity-3.0");
+        int specLevel = WebContainer.getServletContainerSpecLevel();
+
+        /*
+         * Determine which version of appSecurity we are running.
+         */
+        Set<String> installedFeatures = provisionerService.getInstalledFeatures();
+        float appSecurityVers = 0.0f;
+        for (String feature : installedFeatures) {
+            if (feature.startsWith("appSecurity-")) {
+                try {
+                    appSecurityVers = Float.valueOf(feature.substring(12));
+                    break;
+                } catch (NumberFormatException e) {
+                    // Ignore. Not the feature we are looking for.
+                }
+            }
+        }
+
+        /*
+         * Future-proof check. Some of these combinations aren't valid, but
+         * they will not make it this far.
+         */
+        return (appSecurityVers >= 3.0 && specLevel >= WebContainer.SPEC_LEVEL_40) ||
+               (appSecurityVers >= 4.0 && specLevel >= WebContainer.SPEC_LEVEL_50);
```

Comment:
The OR of the conditional will never be triggered, so is there a point in having it?

Reply:
Good catch. I can remove it.